[BUG] Wrong placeholder use in workflow may cause deletion of file

### Description

When using placeholders in a workflow but accidentally misspelling them as `{created_year]` instead of `{created_year}`, it seems difficult/impossible to retrieve the document afterwards. The user gets notified about an invalid token `{` in the filename and an error is shown under file tasks. After that I also restarted the application and performed a reindex.

### Steps to reproduce

1. Create workflow and type `{created_year]` in the assign title field.
2. Upload document which matches the workflow.

### Webserver logs

```bash
[2024-01-06 18:10:30,699] [INFO] [paperless.consumer] Consuming filename.pdf
[2024-01-06 18:10:30,702] [DEBUG] [paperless.consumer] Detected mime type: application/pdf
[2024-01-06 18:10:30,706] [DEBUG] [paperless.consumer] Parser: RasterisedDocumentParser
[2024-01-06 18:10:30,709] [DEBUG] [paperless.consumer] Parsing filename.pdf...
[2024-01-06 18:10:30,932] [DEBUG] [paperless.parsing.tesseract] Calling OCRmyPDF with args: {'input_file': PosixPath('/tmp/paperless/paperless-ngxm_5gu8vf/filename.pdf'), 'output_file': PosixPath('/tmp/paperless/paperless-pajtklp0/archive.pdf'), 'use_threads': True, 'jobs': 2, 'language': 'deu+eng', 'output_type': 'pdfa', 'progress_bar': False, 'color_conversion_strategy': 'RGB', 'skip_text': True, 'clean': True, 'deskew': True, 'rotate_pages': True, 'rotate_pages_threshold': 12.0, 'sidecar': PosixPath('/tmp/paperless/paperless-pajtklp0/sidecar.txt')}
[2024-01-06 18:10:31,366] [DEBUG] [paperless.parsing.tesseract] Incomplete sidecar file: discarding.
[2024-01-06 18:10:31,504] [DEBUG] [paperless.consumer] Generating thumbnail for filename.pdf...
[2024-01-06 18:10:31,507] [DEBUG] [paperless.parsing] Execute: convert -density 300 -scale 500x5000> -alpha remove -strip -auto-orient /tmp/paperless/paperless-pajtklp0/archive.pdf[0] /tmp/paperless/paperless-pajtklp0/convert.webp
[2024-01-06 18:10:32,790] [DEBUG] [paperless.consumer] Saving record to database
[2024-01-06 18:10:32,791] [DEBUG] [paperless.consumer] Creation date from parse_date: 2023-11-30 00:00:00+01:00
[2024-01-06 18:10:32,913] [INFO] [paperless.matching] Document matched WorkflowTrigger 2 from Workflow: workflowname
[2024-01-06 18:10:32,915] [INFO] [paperless.handlers] Applying WorkflowAction 2 from Workflow: workflowname
[2024-01-06 18:10:32,922] [ERROR] [paperless.consumer] The following error occurred while storing document filename.pdf after parsing: unexpected '{' in field name
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/asgiref/sync.py", line 349, in main_wrap
    raise exc_info[1]
  File "/usr/src/paperless/src/documents/consumer.py", line 514, in try_consume_file
    document_consumption_finished.send(
  File "/usr/local/lib/python3.11/site-packages/django/dispatch/dispatcher.py", line 176, in send
    return [
           ^
  File "/usr/local/lib/python3.11/site-packages/django/dispatch/dispatcher.py", line 177, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/paperless/src/documents/signals/handlers.py", line 523, in run_workflow_added
    run_workflow(
  File "/usr/src/paperless/src/documents/signals/handlers.py", line 573, in run_workflow
    document.title = parse_doc_title_w_placeholders(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/paperless/src/documents/consumer.py", line 891, in parse_doc_title_w_placeholders
    return title.format(**formatting).strip()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: unexpected '{' in field name
[2024-01-06 18:10:32,924] [DEBUG] [paperless.parsing.tesseract] Deleting directory /tmp/paperless/paperless-pajtklp0
[2024-01-06 18:11:32,028] [WARNING] [paperless.api] An error occurred listing search results: Document matching query does not exist.
```


### Browser logs

_No response_

### Paperless-ngx version

2.3.0

### Host OS

Debian 12

### Installation method

Docker - official image

### Browser

_No response_

### Configuration changes

_No response_

### Other

_No response_

### Please confirm the following

- [X] I believe this issue is a bug that affects all users of Paperless-ngx, not something specific to my installation.
- [X] I have already searched for relevant existing issues and discussions before opening this report.
- [X] I have updated the title field above with a concise description.

--- a/src-ui/src/app/components/common/edit-dialog/workflow-edit-dialog/workflow-edit-dialog.component.html
+++ b/src-ui/src/app/components/common/edit-dialog/workflow-edit-dialog/workflow-edit-dialog.component.html
@@ -99,7 +99,7 @@
                         <input type="hidden" formControlName="id" />
                         <div class="row">
                           <div class="col">
-                            <pngx-input-text i18n-title title="Assign title" formControlName="assign_title" i18n-hint hint="Can include some placeholders, see <a target='_blank' href='https://docs.paperless-ngx.com/usage/#workflows'>documentation</a>." [error]="error?.assign_title"></pngx-input-text>
+                            <pngx-input-text i18n-title title="Assign title" formControlName="assign_title" i18n-hint hint="Can include some placeholders, see <a target='_blank' href='https://docs.paperless-ngx.com/usage/#workflows'>documentation</a>." [error]="error?.actions?.[i]?.assign_title"></pngx-input-text>
                             <pngx-input-tags [allowCreate]="false" i18n-title title="Assign tags" formControlName="assign_tags"></pngx-input-tags>
                             <pngx-input-select i18n-title title="Assign document type" [items]="documentTypes" [allowNull]="true" formControlName="assign_document_type"></pngx-input-select>
                             <pngx-input-select i18n-title title="Assign correspondent" [items]="correspondents" [allowNull]="true" formControlName="assign_correspondent"></pngx-input-select>

--- a/src/documents/consumer.py
+++ b/src/documents/consumer.py
@@ -730,9 +730,9 @@ class Consumer(LoggingMixin):
         if self.override_title is not None:
             try:
                 title = self._parse_title_placeholders(self.override_title)
-            except Exception:
+            except Exception as e:
                 self.log.error(
-                    f"Error occurred parsing title override '{self.override_title}', falling back to original",
+                    f"Error occurred parsing title override '{self.override_title}', falling back to original. Exception: {e}",
                 )
 
         document = Document.objects.create(

--- a/src/documents/consumer.py
+++ b/src/documents/consumer.py
@@ -726,12 +726,17 @@ class Consumer(LoggingMixin):
 
         storage_type = Document.STORAGE_TYPE_UNENCRYPTED
 
+        title = file_info.title[:127]
+        if self.override_title is not None:
+            try:
+                title = self._parse_title_placeholders(self.override_title)
+            except Exception:
+                self.log.error(
+                    f"Error occurred parsing title override '{self.override_title}', falling back to original",
+                )
+
         document = Document.objects.create(
-            title=(
-                self._parse_title_placeholders(self.override_title)
-                if self.override_title is not None
-                else file_info.title
-            )[:127],
+            title=title,
             content=text,
             mime_type=mime_type,
             checksum=hashlib.md5(self.working_copy.read_bytes()).hexdigest(),

--- a/src/documents/serialisers.py
+++ b/src/documents/serialisers.py
@@ -1414,9 +1414,9 @@ class WorkflowActionSerializer(serializers.ModelSerializer):
                         created_day="",
                         created_time="",
                     )
-                except ValueError as e:
+                except (ValueError, KeyError) as e:
                     raise serializers.ValidationError(
-                        {"assign_title": f"Invalid f-string detected: {e.args[0]}"},
+                        {"assign_title": f'Invalid f-string detected: "{e.args[0]}"'},
                     )
 
         return attrs

--- a/src/documents/serialisers.py
+++ b/src/documents/serialisers.py
@@ -1386,12 +1386,38 @@ class WorkflowActionSerializer(serializers.ModelSerializer):
 
     def validate(self, attrs):
         # Empty strings treated as None to avoid unexpected behavior
-        if (
-            "assign_title" in attrs
-            and attrs["assign_title"] is not None
-            and len(attrs["assign_title"]) == 0
-        ):
-            attrs["assign_title"] = None
+        if "assign_title" in attrs:
+            if attrs["assign_title"] is not None and len(attrs["assign_title"]) == 0:
+                attrs["assign_title"] = None
+            else:
+                try:
+                    # test against all placeholders, see consumer.py `parse_doc_title_w_placeholders`
+                    attrs["assign_title"].format(
+                        correspondent="",
+                        document_type="",
+                        added="",
+                        added_year="",
+                        added_year_short="",
+                        added_month="",
+                        added_month_name="",
+                        added_month_name_short="",
+                        added_day="",
+                        added_time="",
+                        owner_username="",
+                        original_filename="",
+                        created="",
+                        created_year="",
+                        created_year_short="",
+                        created_month="",
+                        created_month_name="",
+                        created_month_name_short="",
+                        created_day="",
+                        created_time="",
+                    )
+                except ValueError as e:
+                    raise serializers.ValidationError(
+                        {"assign_title": f"Invalid f-string detected: {e.args[0]}"},
+                    )
 
         return attrs
 

--- a/src/documents/signals/handlers.py
+++ b/src/documents/signals/handlers.py
@@ -570,19 +570,27 @@ def run_workflow(
                     document.owner = action.assign_owner
 
                 if action.assign_title is not None:
-                    document.title = parse_doc_title_w_placeholders(
-                        action.assign_title,
-                        document.correspondent.name
-                        if document.correspondent is not None
-                        else "",
-                        document.document_type.name
-                        if document.document_type is not None
-                        else "",
-                        document.owner.username if document.owner is not None else "",
-                        timezone.localtime(document.added),
-                        document.original_filename,
-                        timezone.localtime(document.created),
-                    )
+                    try:
+                        document.title = parse_doc_title_w_placeholders(
+                            action.assign_title,
+                            document.correspondent.name
+                            if document.correspondent is not None
+                            else "",
+                            document.document_type.name
+                            if document.document_type is not None
+                            else "",
+                            document.owner.username
+                            if document.owner is not None
+                            else "",
+                            document.added,
+                            document.original_filename,
+                            document.created,
+                        )
+                    except Exception:
+                        logger.error(
+                            f"Error occurred parsing title assignment '{action.assign_title}', falling back to original",
+                            extra={"group": logging_group},
+                        )
 
                 if (
                     action.assign_view_users is not None

--- a/src/documents/signals/handlers.py
+++ b/src/documents/signals/handlers.py
@@ -587,7 +587,7 @@ def run_workflow(
                             document.created,
                         )
                     except Exception:
-                        logger.error(
+                        logger.exception(
                             f"Error occurred parsing title assignment '{action.assign_title}', falling back to original",
                             extra={"group": logging_group},
                         )

--- a/src/documents/tests/test_api_workflows.py
+++ b/src/documents/tests/test_api_workflows.py
@@ -248,6 +248,45 @@ class TestApiWorkflows(DirectoriesMixin, APITestCase):
 
         self.assertEqual(WorkflowTrigger.objects.count(), 1)
 
+    def test_api_create_invalid_assign_title(self):
+        """
+        GIVEN:
+            - API request to create a workflow
+            - Invalid f-string for assign_title
+        WHEN:
+            - API is called
+        THEN:
+            - Correct HTTP 400 response
+            - No objects are created
+        """
+        response = self.client.post(
+            self.ENDPOINT,
+            json.dumps(
+                {
+                    "name": "Workflow 1",
+                    "order": 1,
+                    "triggers": [
+                        {
+                            "type": WorkflowTrigger.WorkflowTriggerType.DOCUMENT_UPDATED,
+                        },
+                    ],
+                    "actions": [
+                        {
+                            "assign_title": "{created_year]",
+                        },
+                    ],
+                },
+            ),
+            content_type="application/json",
+        )
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertIn(
+            "Invalid f-string detected",
+            response.data["actions"][0]["assign_title"][0],
+        )
+
+        self.assertEqual(Workflow.objects.count(), 1)
+
     def test_api_create_workflow_trigger_action_empty_fields(self):
         """
         GIVEN:

--- a/src/documents/tests/test_consumer.py
+++ b/src/documents/tests/test_consumer.py
@@ -423,6 +423,16 @@ class TestConsumer(DirectoriesMixin, FileSystemAssertsMixin, TestCase):
         self.assertEqual(document.title, "Override Title")
         self._assert_first_last_send_progress()
 
+    def testOverrideTitleInvalidPlaceholders(self):
+        with self.assertLogs("paperless.consumer", level="ERROR") as cm:
+            document = self.consumer.try_consume_file(
+                self.get_test_file(),
+                override_title="Override {correspondent]",
+            )
+            self.assertEqual(document.title, "sample")
+            expected_str = "Error occurred parsing title override 'Override {correspondent]', falling back to original"
+            self.assertIn(expected_str, cm.output[0])
+
     def testOverrideCorrespondent(self):
         c = Correspondent.objects.create(name="test")
 

--- a/src/documents/tests/test_workflows.py
+++ b/src/documents/tests/test_workflows.py
@@ -966,6 +966,50 @@ class TestWorkflows(DirectoriesMixin, FileSystemAssertsMixin, APITestCase):
             expected_str = f"Document correspondent {doc.correspondent} does not match {trigger.filter_has_correspondent}"
             self.assertIn(expected_str, cm.output[1])
 
+    def test_document_added_invalid_title_placeholders(self):
+        """
+        GIVEN:
+            - Existing workflow with added trigger type
+            - Assign title field has an error
+        WHEN:
+            - File that matches is added
+        THEN:
+            - Title is not updated, error is output
+        """
+        trigger = WorkflowTrigger.objects.create(
+            type=WorkflowTrigger.WorkflowTriggerType.DOCUMENT_ADDED,
+            filter_filename="*sample*",
+        )
+        action = WorkflowAction.objects.create(
+            assign_title="Doc {created_year]",
+        )
+        w = Workflow.objects.create(
+            name="Workflow 1",
+            order=0,
+        )
+        w.triggers.add(trigger)
+        w.actions.add(action)
+        w.save()
+
+        now = timezone.localtime(timezone.now())
+        created = now - timedelta(weeks=520)
+        doc = Document.objects.create(
+            original_filename="sample.pdf",
+            title="sample test",
+            content="Hello world bar",
+            created=created,
+        )
+
+        with self.assertLogs("paperless.handlers", level="ERROR") as cm:
+            document_consumption_finished.send(
+                sender=self.__class__,
+                document=doc,
+            )
+            expected_str = f"Error occurred parsing title assignment '{action.assign_title}', falling back to original"
+            self.assertIn(expected_str, cm.output[0])
+
+        self.assertEqual(doc.title, "sample test")
+
     def test_document_updated_workflow(self):
         trigger = WorkflowTrigger.objects.create(
             type=WorkflowTrigger.WorkflowTriggerType.DOCUMENT_UPDATED,

